python/thunder/imgprocessing/transformation.py

```diff
@@ -78,3 +241,98 @@ def apply(self, im):
+        self.dims = np.array(dims)
+        self.ndim = len(dims)
+        if center is None:
+            center = (self.dims - 1) / 2.0
+        self.center = center
+        self.homo_points = np.ones((self.ndim + 1,) + dims)
+        self.homo_points[:-1, ...] = np.array(np.mgrid[[slice(p) for p in self.dims]])
+        self.raw_points = self.homo_points.copy()
+        self.world_to_index_tfm = transformationMatrix(self.center, np.zeros(self.ndim))
+        self.index_to_world_tfm = transformationMatrix(-self.center, np.zeros(self.ndim))
+        self.homo_points = self.transform_grid_world(self.index_to_world_tfm)
+        self.index_points = self.transform_grid_world(self.world_to_index_tfm)
+
+    def transform_grid_world(self, A):
+        """Get the grid of points in world space after applying the given affine transform."""
+        return np.tensordot(A.T, self.homo_points, axes=(0, 0))
+
+    def transform_grid(self, A):
+        """Get the grid of points in index space after applying the given affine transform
+           in world space.
+
+        Args:
+            A: 4x4 Affine transformation matrix
+        Returns:
+            y: 4 x dims[0] x dims[1] x dims[2] matrix containing the grid
+        """
+        return self.transform_grid_world(np.dot(self.world_to_index_tfm, A))
+
+    def transform_vol(self, vol, A, **kwargs):
+        from scipy.ndimage import map_coordinates
+        new_grid = self.transform_grid(A)[:-1]
+        transformed_vol = map_coordinates(vol, new_grid, cval=0.0, **kwargs)
+        return transformed_vol
+
+
+def transformationMatrix(shift, rot=None):
+    """Create an affine transformation matrix
+
+    Parameters
+    ----------
+    shift : array, shape (ndim,)
+        translations along each dimension
+    rot : scalar or array with shape (ndim, )
+
+    Returns
+    -------
+    A : array, shape (ndims + 1, ndims + 1)
+        transformation matrix that shifts and rotates a set of points in homogeneous coordinates
+    """
+
+    ndim = len(shift)
+    if rot is None:
+        rot = np.zeros(ndim)
+    else:
+        rot = np.atleast_1d(rot)
+    c = np.cos(rot)
+    s = np.sin(rot)
+    trans = np.eye(ndim + 1)
+    trans[:-1, -1] = shift
+    xrot = np.array(
+            [[c[0], -s[0],  0,    0],
+            [s[0],  c[0],  0,    0],
+            [0,     0,     1,    0],
+            [0,     0,     0,    1]])
+    if ndim == 2 or np.size(rot) == 1:
+        A = np.dot(trans, xrot[:ndim + 1, :ndim + 1])
+    else:
+        yrot = [[c[1],  0,     s[1], 0,],
+                [0,     1,     0,    0],
+                [-s[1], 0,     c[1], 0],
+                [0,     0,     0,    1]]
+        zrot = [[1,     0,     0,    0],
+                [0,     c[2], -s[2], 0],
+                [0,     s[2],  c[2], 0],
+                [0,     0,     0,    1]]
+        A = np.dot(trans, np.dot(xrot, np.dot(yrot, zrot)))
+    return A
+
+# Dict of valid types of Transformations used by Lucas-Kanade
```

Comment:
Probably don't want to put this here, as this whole module isn't specific to the LK algorithm.